### PR TITLE
Fix Failed to connect to elasticsearch-discovery headless service

### LINKED ISSUE #9243: [stable/elasticsearch] Failed to connect to elasticsearch-discovery headless service / Failed to resolve host
**Is this a request for help?**:
Yes
---

**Is this a BUG REPORT or FEATURE REQUEST?**:
BUG REPORT

**Version of Helm and Kubernetes**:
K8s v 1.11.4
helm v2.10 

**Which chart**:
stable/elasticsearch

**What happened**:
I've been trying to get Elasticsearch running on K8s using this chart for Elasticsearch v6.4.3

However the master pod fail with inability to connect to elasticsearch-discovery headless service.

master-0 (the only master that attempts to start) fails with:

> [2018-11-13T20:07:57,315][WARN ][o.e.d.z.ZenDiscovery     ] [elasticsearch-data-0] not enough master nodes discovered during pinging (found [[]], but needed [2]), pinging again
> [2018-11-13T20:07:57,316][WARN ][o.e.d.z.UnicastZenPing   ] [elasticsearch-data-0] failed to resolve host [elasticsearch-discovery]

This is because, according to [documentation](https://github.com/kubernetes/dns/blob/00f97cb14c316087cd9a01af4954c4ff0b2756ba/docs/specification.md):

`If there are no ready endpoints for the headless Service, the answer should be NXDOMAIN`

**What you expected to happen**:
The master must have the ability to connect to the others master nodes without the use of a headless service.

**How to reproduce it**:
`helm install <this-char>` ... with all defaults

### LINKED ISSUE #9343: [stable/elasticsearch]: Not worked with elasticsearch version 5.4 5.2.2 5.6.13 etc.
**Is this a request for help?**:
Yes
---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
BUG REPORT

**Version of Helm and Kubernetes**:
helm: v2.11.0
kubernetes: v1.11.1

**Which chart**:
stable/elasticsearch

**What happened**:
except for appVersion and image, everything else is exactly the same.Then 6.4.2 works well, but 5.2.2/5.4/5.6.13 didn't work!

**What you expected to happen**:
work well

**How to reproduce it** (as minimally and precisely as possible):
values.yaml diff with different versions just like:
-appVersion: "6.4.2"
+appVersion: "5.4"

-repository: "docker.elastic.co/elasticsearch/elasticsearch-oss"
-tag: "6.4.2"
+repository: "docker.elastic.co/elasticsearch/elasticsearch"
+tag: "5.4"

### PR DESCRIPTION
**Pull request by JasonJPu**
Signed-off-by: JasonJPu <cub1@live.ca>

#### Which issue this PR fixes
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
  - fixes #9243 
  - fixes #9343

#### Checklist
[Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields.]
- [x] [DCO](https://www.helm.sh/blog/helm-dco/index.html) signed
- [x] Chart Version bumped
- [x] Variables are documented in the README.md

### CODE REVIEW

**Comment by anapsix** [comment]
@JasonJPu should this be done to `stable/elasticsearch`, rather than `incubator/elasticsearch`? 
Afaik, incubator version is deprecated.

**Comment by JasonJPu** [comment]
@anapsix You're right, I made the change

File: stable/elasticsearch/templates/client-svc.yaml [hunk 1]
```diff
@@ -10,6 +10,7 @@ metadata:
   name: {{ template "elasticsearch.client.fullname" . }}
 {{- if .Values.client.serviceAnnotations }}
   annotations:
+    service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"
```
Comment (rendhalver) [inline stable/elasticsearch/templates/client-svc.yaml:13]: How does alloying unready pods into the service fix the issue?

**Review by rendhalver — CHANGES_REQUESTED** [review]

File: stable/elasticsearch/templates/client-svc.yaml (hunk 1 above)
Comment (JasonJPu) [reply stable/elasticsearch/templates/client-svc.yaml:13]: The pods are unready since they can’t find the service, and the service fails because it doesn’t find any pods. It’s like a circular loop

**Comment by anapsix** [comment]
@rendhalver, without this the following messages are logged (see https://github.com/helm/charts/issues/9243#issuecomment-446671096 for more details)
```
[2018-12-12T17:20:58,916][WARN ][o.e.d.z.UnicastZenPing   ] [broken-record-elasticsearch-master-0] failed to resolve host [broken-record-elasticsearch-discovery]
java.net.UnknownHostException: broken-record-elasticsearch-discovery
	at java.net.InetAddress.getAllByName0(InetAddress.java:1281) ~[?:1.8.0_191]
	at java.net.InetAddress.getAllByName(InetAddress.java:1193) ~[?:1.8.0_191]
	at java.net.InetAddress.getAllByName(InetAddress.java:1127) ~[?:1.8.0_191]
	at org.elasticsearch.transport.TcpTransport.parse(TcpTransport.java:911) ~[elasticsearch-5.6.14.jar:5.6.14]
	at org.elasticsearch.transport.TcpTransport.addressesFromString(TcpTransport.java:866) ~[elasticsearch-5.6.14.jar:5.6.14]
	at org.elasticsearch.transport.TransportService.addressesFromString(TransportService.java:709) ~[elasticsearch-5.6.14.jar:5.6.14]
	at org.elasticsearch.discovery.zen.UnicastZenPing.lambda$null$0(UnicastZenPing.java:212) ~[elasticsearch-5.6.14.jar:5.6.14]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_191]
	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:576) [elasticsearch-5.6.14.jar:5.6.14]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_191]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_191]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_191]
```

**Comment by monotek** [comment]
Is this maybe only an issue when the 5.6.x images are used?

P.S.: Please resolve merge conflicts.

**Comment by anapsix** [comment]
Yes @monotek, this is addresses an issue with ES 5.6.x. ES 6+ versions not affected, in my tests.

**Comment by monotek** [comment]
OK, in this case the "service.alpha.kubernetes.io/tolerate-unready-endpoints: "true"" annotation should imho only be used if a 5.6.x image is used.

**Comment by anapsix** [comment]
Argh.. I'm tracking another issue with which is ES 5.6 specific.. and have mistook this one for it.
I take back my previous comment, though 5.6 is definitely affected, I don't recall if 6+ is unaffected @monotek

**Comment by anapsix** [comment]
Here's an example of this affecting ES 6+ [pull-charts-e2e]
(https://gubernator.k8s.io/build/kubernetes-jenkins/pr-logs/pull/helm_charts/9124/pull-charts-e2e/17956/?log#log) for unrelated PR
/cc @monotek

**Comment by erictcgs** [comment]
Hi @monotek , @JasonJPu , @anapsix - Thanks for starting this - I'm deploying 6.1 now and was seeing this error.  

2 additions that were necessary for me (at least for 6.1) beyond your changes:
- in `client-svc.yml` you include the `annotations:` and tolerate-ready-endpoints in the if conditional of whether annotations are specified in the user's values.yaml, but these should be specified regardless of whether the user has additional annotations, correct?
- In 6.1 at least, k8s was getting into a chicken-and-egg state where the master statefulset started pod-0, which came up as unready and started waiting for other masters to join and form quorum before it would report ready, which prevented k8s from continuing the statefulset rollout.  I added `podManagementPolicy: "Parallel"` to the statefulset spec to allow all masters to start simultaneously and begin joining one another

File: stable/elasticsearch/templates/client-svc.yaml (hunk 1 above)
Comment (rendhalver) [reply stable/elasticsearch/templates/client-svc.yaml:13]: That doesn't answer the question.
We should fix the health check instead of ignore it.

Comment (anapsix) [reply stable/elasticsearch/templates/client-svc.yaml:13]: Master's readinessProbe is `httpGet` of `/_cluster/health?local=true` on port 9200
If in certain version of ES healthcheck success or failure depends on masters discovering other masters adding `tolerate-unready-endpoints: true` to headless service is justified. There is nothing that can be fixed about healthcheck @rendhalver. Please see https://github.com/helm/charts/pull/9910#issuecomment-451515387 as well.

**Comment by anapsix** [comment]
### TL;DR

For `5.6.14`, use extra cluster config `cluster.config."xpack.security.enabled": false`

### Details
Just provisioned 5.6.14 cluster with Elasticsearch chart v1.19.1, and still getting master's statefulset rollout getting stuck.
Master-0 logs continuously spammed with following errors:
```
[2019-02-15T13:08:50,132][WARN ][o.e.d.z.ZenDiscovery     ] [myorg-elasticsearch-master-0] not enough master nodes discovered during pinging (found [[Candidate{node={myorg-elasticsearch-master-0}{BgrsK3gATWyCRP4dZqQ2bA}{E5EllcnnRWGl8ZeqbhXhKA}{192.168.7.108}{192.168.7.108:9300}{ml.max_open_jobs=10, ml.enabled=true}, clusterStateVersion=-1}]], but needed [2]), pinging again
[2019-02-15T13:08:50,133][WARN ][o.e.d.z.UnicastZenPing   ] [myorg-elasticsearch-master-0] failed to resolve host [myorg-elasticsearch-discovery]
java.net.UnknownHostException: myorg-elasticsearch-discovery
  at java.net.InetAddress.getAllByName0(InetAddress.java:1281) ~[?:1.8.0_191]
  at java.net.InetAddress.getAllByName(InetAddress.java:1193) ~[?:1.8.0_191]
  at java.net.InetAddress.getAllByName(InetAddress.java:1127) ~[?:1.8.0_191]
  at org.elasticsearch.transport.TcpTransport.parse(TcpTransport.java:911) ~[elasticsearch-5.6.14.jar:5.6.14]
  at org.elasticsearch.transport.TcpTransport.addressesFromString(TcpTransport.java:866) ~[elasticsearch-5.6.14.jar:5.6.14]
  at org.elasticsearch.transport.TransportService.addressesFromString(TransportService.java:709) ~[elasticsearch-5.6.14.jar:5.6.14]
  at org.elasticsearch.discovery.zen.UnicastZenPing.lambda$null$0(UnicastZenPing.java:212) ~[elasticsearch-5.6.14.jar:5.6.14]
  at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[?:1.8.0_191]
  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:576) [elasticsearch-5.6.14.jar:5.6.14]
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_191]
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_191]
  at java.lang.Thread.run(Thread.java:748) [?:1.8.0_191]
```

Manually setting `spec.publishNotReadyAddresses: true` on master's headless service `..-elasticsearch-discovery`, changed log output to
```
[2019-02-15T13:23:20,775][WARN ][o.e.d.z.ZenDiscovery     ] [myorg-elasticsearch-master-0] not enough master nodes discovered during pinging (found [[Candidate{node={myorg-elasticsearch-master-0}{R9ckNBmuQJaUlshnR5XPzg}{1QGEbUTWSBG7to6_QAlebw}{192.168.4.95}{192.168.4.95:9300}{ml.max_open_jobs=10, ml.enabled=true}, clusterStateVersion=-1}]], but needed [2]), pinging again
```

Still, only master's statefulset rollout stopped at a single master pod, since healthcheck continued to return error:
```
sh-4.2$ curl 127.0.0.1:9200/_cluster/health?local=true
{"error":{"root_cause":[{"type":"security_exception","reason":"missing authentication token for REST request [/_cluster/health?local=true]","header":{"WWW-Authenticate":"Basic realm=\"security\" charset=\"UTF-8\""}}],"type":"security_exception","reason":"missing authentication token for REST request [/_cluster/health?local=true]","header":{"WWW-Authenticate":"Basic realm=\"security\" charset=\"UTF-8\""}},"status":401}
```

So the problem lies, not in unresolvable master's discovery service hostname, but in healthcheck (see output above, specifically "security_exception").

> NOTE: I'm using `docker.elastic.co/elasticsearch/elasticsearch:5.6.14` image

Starting fresh with no changes to master's headless discovery service spec or annotations, and making a single addition of `cluster.config."xpack.security.enabled": false`, explicitly disables security, allowing healthcheck to succeed. Even though few of `ZenDiscovery` errors still appear in logs, while master node passes readiness check.

Unfortunately, the condition that meant to disable X-Pack explicitly in [configmap.yaml#L45][1] does not work as expected.

## Conclusion
Thus, this Pull Request won't fix the problem, and only addresses somewhat annoying error messaging.. And indeed, there is no need to discover unhealthy masters.
Another Pull Request is needed to fix the issue in [configmap.yaml#L45][1]


[Link Reference]::
[1]: https://github.com/helm/charts/blob/master/stable/elasticsearch/templates/configmap.yaml#L45
